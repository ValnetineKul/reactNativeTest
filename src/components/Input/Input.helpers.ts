export const inputLabelPositionYAnimationInterpolation = {
  inputRange: [0, 1],
  outputRange: [0, -18],
};

export const inputLabelScaleAnimationInterpolation = {
  inputRange: [0, 1],
  outputRange: [1.1, 1],
};
