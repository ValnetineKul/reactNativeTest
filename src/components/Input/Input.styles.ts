import { StyleSheet } from "react-native";
import { COLOR_TEXT_GRAY } from "../../theme";

export const styles = StyleSheet.create({
  inputContainer: {
    borderWidth: 1,
    borderColor: COLOR_TEXT_GRAY,
    borderRadius: 4,
    position: "relative",
  },
  input: {
    height: 34,
    paddingLeft: 40,
  },
  startAdorment: {
    position: "absolute",
    left: 10,
    top: 6,
  },
});
