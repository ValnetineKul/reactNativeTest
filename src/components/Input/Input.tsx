import React, { ReactElement, forwardRef, useEffect, useRef, useState } from "react";
import {
  KeyboardTypeOptions,
  View,
  TextInput,
  ViewStyle,
  NativeSyntheticEvent,
  TextInputFocusEventData,
  Animated,
  TouchableOpacity,
} from "react-native";
import { Typography } from "../Typography";
import { styles } from "./Input.styles";
import {
  inputLabelPositionYAnimationInterpolation,
  inputLabelScaleAnimationInterpolation,
} from "./Input.helpers";
import { Loader } from "../Loader";

type InputProps = {
  onChange?: (text: string) => void;
  onBlur?: (event: NativeSyntheticEvent<TextInputFocusEventData>) => void;
  onFocus?: (event: NativeSyntheticEvent<TextInputFocusEventData>) => void;
  onStartAdormentPress?: () => void;
  label?: string;
  value?: string;
  startAdorment?: ReactElement;
  endAdorment?: ReactElement;
  keyboardType?: KeyboardTypeOptions;
  placeHolder?: string;
  style?: ViewStyle | ViewStyle[];
  error?: string;
  type?: string;
  disabled?: boolean;
  isLoading?: boolean;
};

export const Input = forwardRef<TextInput, InputProps>(
  (
    {
      placeHolder = "",
      onChange,
      onBlur,
      onFocus,
      onStartAdormentPress,
      label,
      value,
      startAdorment,
      endAdorment,
      keyboardType = "default",
      error,
      style,
      type,
      disabled,
      isLoading,
    },
    ref
  ) => {
    const [isFocused, setFocused] = useState(false);
    const inputLabelAnimationValue = useRef(new Animated.Value(0)).current;

    const handleBlur = (event: NativeSyntheticEvent<TextInputFocusEventData>) => {
      onBlur?.(event);
      setFocused(false);
    };

    const handleFocus = (event: NativeSyntheticEvent<TextInputFocusEventData>) => {
      onFocus?.(event);
      setFocused(true);
    };

    useEffect(() => {
      Animated.timing(inputLabelAnimationValue, {
        toValue: Number(isFocused) || Number(!!value?.length),
        duration: 200,
        useNativeDriver: true,
      }).start();
    }, [inputLabelAnimationValue, isFocused, value?.length]);

    return (
      <View style={[styles.inputContainer, { ...(error && styles.error) }, style]}>
        {startAdorment && (
          <TouchableOpacity onPress={onStartAdormentPress} style={styles.startAdorment}>
            {startAdorment}
          </TouchableOpacity>
        )}
        {label && (
          <Animated.View
            style={[
              styles.labelContainer,
              {
                transform: [
                  {
                    translateY: inputLabelAnimationValue.interpolate(
                      inputLabelPositionYAnimationInterpolation
                    ),
                  },
                  {
                    scale: inputLabelAnimationValue.interpolate(inputLabelScaleAnimationInterpolation),
                  },
                ],
              },
            ]}
          >
            <Typography variant="caption">{label}</Typography>
          </Animated.View>
        )}
        <TextInput
          onChangeText={onChange}
          value={value}
          placeholder={placeHolder}
          keyboardType={keyboardType}
          style={[styles.input, startAdorment && styles.inputStartAdormentPadding]}
          secureTextEntry={type === "password"}
          onBlur={handleBlur}
          onFocus={handleFocus}
          editable={!disabled}
          selectTextOnFocus={!disabled}
          ref={ref}
        />
        {isLoading && <Loader style={[styles.loader]} color="blue" />}
        {endAdorment}
        {!!error && (
          <Typography style={[styles.errorText]} variant="caption" color="error">
            {error}
          </Typography>
        )}
      </View>
    );
  }
);
