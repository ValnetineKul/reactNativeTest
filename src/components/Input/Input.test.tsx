import React from "react";
import renderer from 'react-test-renderer';
import { render, screen } from "@testing-library/react-native";
import { Input } from "./Input";

describe("Input tests", () => {
  it("renders the correct text", () => {
    render(<Input value="Hello world" />);
    expect(screen.getByDisplayValue("Hello world")).toBeTruthy();
  });

  it('renders input correctly', () => {
    const tree = renderer.create(<Input />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
