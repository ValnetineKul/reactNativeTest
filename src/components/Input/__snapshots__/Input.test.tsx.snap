// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests renders input correctly 1`] = `
<View
  style={
    [
      {
        "borderColor": "#8F8F8F",
        "borderRadius": 4,
        "borderWidth": 1,
        "position": "relative",
        "width": "100%",
      },
      {},
      undefined,
    ]
  }
>
  <TextInput
    editable={true}
    keyboardType="default"
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder=""
    secureTextEntry={false}
    selectTextOnFocus={true}
    style={
      [
        {
          "height": 34,
          "paddingLeft": 16,
        },
        undefined,
      ]
    }
  />
</View>
`;
